File: Flink/VoipStream/src/main/java/VoipStream/CT24.java
package VoipStream;

import org.apache.flink.api.common.functions.RichFlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple6;
import org.apache.flink.api.java.tuple.Tuple7;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.util.Collector;
import org.slf4j.Logger;
import common.CallDetailRecord;
import common.Constants;
import common.ODTDBloomFilter;
import common.ScorerMap;
import util.Log;

public class CT24 extends RichFlatMapFunction<
        Tuple7<Long, String, String, Long, Boolean, CallDetailRecord, Double>,
        Tuple6<Integer, Long, String, Long, Double, CallDetailRecord>> {
    private static final Logger LOG = Log.get(CT24.class);

    private ODTDBloomFilter filter;

    @Override
    public void open(Configuration parameters) {

        int numElements = Constants.CT24_NUM_ELEMENTS;
        int bucketsPerElement = Constants.CT24_BUCKETS_PER_ELEMENT;
        int bucketsPerWord = Constants.CT24_BUCKETS_PER_WORD;
        double beta = Constants.CT24_BETA;
        filter = new ODTDBloomFilter(numElements, bucketsPerElement, beta, bucketsPerWord);
    }

    @Override
    public void flatMap(Tuple7<Long, String, String, Long, Boolean, CallDetailRecord, Double> tuple, Collector<Tuple6<Integer, Long, String, Long, Double, CallDetailRecord>> collector) {
        LOG.debug("tuple in: {}", tuple);

        long timestamp = tuple.f0;
        CallDetailRecord cdr = tuple.f5;

        boolean newCallee = tuple.f4;
        if (cdr.callEstablished && newCallee) {
            String caller = tuple.f1;
            filter.add(caller, cdr.callDuration, cdr.answerTimestamp);
            double callTime = filter.estimateCount(caller, cdr.answerTimestamp);

            Tuple6<Integer, Long, String, Long, Double, CallDetailRecord> out = new Tuple6<>(ScorerMap.CT24, timestamp, caller, cdr.answerTimestamp, callTime, cdr);
            collector.collect(out);
            LOG.debug("tuple out: {}", out);
        }
    }
}

File: WindFlow/VoipStream/lib/voip_stream/ct24.cpp
#include "constants.hpp"
#include "ct24.hpp"
#include "util/log.hpp"

namespace voip_stream {

CT24::CT24()
    : filter_(CT24_NUM_ELEMENTS, CT24_BUCKETS_PER_ELEMENT, CT24_BETA, CT24_BUCKETS_PER_WORD)
{}

void CT24::operator ()(const Dispatcher::Tuple &tuple, wf::Shipper<FilterTuple> &shipper, wf::RuntimeContext &rc)
{
    DEBUG_LOG("ct24::tuple " << tuple);

    if (tuple.cdr.call_established && tuple.new_callee) {
        filter_.add(tuple.cdr.calling_number, tuple.cdr.call_duration, tuple.cdr.answer_timestamp);
        double call_time = filter_.estimate_count(tuple.cdr.calling_number, tuple.cdr.answer_timestamp);

        FilterTuple result;
        result.ts = tuple.ts;
        result.source = SOURCE_CT24;
        result.rate = call_time;
        result.cdr = tuple.cdr;
        shipper.push(result);
    }
}

}
